hotel_restaurant/models/hotel_restaurant.py

```diff
@@ -0,0 +1,690 @@
+            sale.amount_subtotal = sum(
+                line.price_subtotal for line in sale.order_list_ids
+            )
+            sale.amount_total = 0.0
+            if sale.amount_subtotal:
+                sale.amount_total = (
+                    sale.amount_subtotal + (sale.amount_subtotal * sale.tax) / 100
+                )
+
+    def done_cancel(self):
+        """
+        This method is used to change the state
+        to cancel of the hotel restaurant order
+        ----------------------------------------
+        @param self: object pointer
+        """
+        self.write({"state": "cancel"})
+
+    def set_to_draft(self):
+        """
+        This method is used to change the state
+        to draft of the hotel restaurant order
+        ----------------------------------------
+        @param self: object pointer
+        """
+        self.write({"state": "draft"})
+
+    def generate_kot(self):
+        """
+        This method create new record for hotel restaurant order list.
+        @param self: The object pointer
+        @return: new record set for hotel restaurant order list.
+        """
+        res = []
+        order_tickets_obj = self.env["hotel.restaurant.kitchen.order.tickets"]
+        restaurant_order_list_obj = self.env["hotel.restaurant.order.list"]
+        for order in self:
+            if not order.order_list_ids:
+                raise ValidationError(_("Please Give an Order"))
+            if not order.table_nos_ids:
+                raise ValidationError(_("Please Assign a Table"))
+            table_ids = order.table_nos_ids.ids
+            kot_data = order_tickets_obj.create(
+                {
+                    "orderno": order.order_no,
+                    "kot_date": order.o_date,
+                    "room_no": order.room_id.name,
+                    "w_name": order.waiter_id.name,
+                    "table_nos_ids": [(6, 0, table_ids)],
+                }
+            )
+
+            for order_line in order.order_list_ids:
+                o_line = {
+                    "kot_order_id": kot_data.id,
+                    "menucard_id": order_line.menucard_id.id,
+                    "item_qty": order_line.item_qty,
+                    "item_rate": order_line.item_rate,
+                }
+                restaurant_order_list_obj.create(o_line)
+                res.append(order_line.id)
+            order.update(
+                {
+                    "kitchen": kot_data.id,
+                    "rest_item_id": [(6, 0, res)],
+                    "state": "order",
+                }
+            )
+        return True
+
+    order_no = fields.Char("Order Number", readonly=True)
+    o_date = fields.Datetime(
+        "Order Date", required=True, default=lambda self: fields.Datetime.now()
+    )
+    room_id = fields.Many2one("product.product", "Room No")
+    folio_id = fields.Many2one("hotel.folio", "Folio No")
+    waiter_id = fields.Many2one("res.partner", "Waiter Name")
```

Comment:
Make sure to make consistent field type,
At one place Waiter Name is Char field and here m2o